### PR TITLE
 ISO files: add a new modules iso_customize #5048

### LINKED ISSUE #5048: Add support for customizing ISO ISO in community.general.iso_customize
### Summary

1. New feature request for customizing iso:
   1) add files to ISO
   2) delete files from ISO
  

### Issue Type

Feature Idea

### Component Name

iso_customize 
### Additional Information

```yaml (paste below)

```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by ZouYuhua**
##### SUMMARY
Fixes #5048 
1. New feature request for customizing iso:
        add files to ISO
        delete files from ISO
     

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
iso_customize

##### ADDITIONAL INFORMATION
Add a new module named "iso_customize" With the module, we can easily add unattend file to iso for unattend installation

### CODE REVIEW

File: changelogs/fragments/5048-iso-customize-getfiles.yml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+minor_changes:
+  - iso_customize - add new iso_customize.py module file in files folder
+  - iso_get_files - add new iso_get_files.py module file in files folder
```
Comment (felixfontein) [inline changelogs/fragments/5048-iso-customize-getfiles.yml:3]: Please remove this. New modules should not have a changelog fragment. (Generally see `CONTRIBUTING.md` for information on how to add new modules.)

File: tests/integration/targets/iso_customize/files/grub.cfg [hunk 2]
```diff
@@ -0,0 +1,35 @@
+set timeout=30
```
Comment (felixfontein) [inline tests/integration/targets/iso_customize/files/grub.cfg:1]: You need an `aliases file in `tests/integration/targets/iso_customize/` for tests to actually run in CI.

File: plugins/modules/files/iso_get_files.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+module: iso_get_files
+short_description: Fetch files from ISO
+description:
+    - This module is used to get specific files from ISO.
+author:
+    - Yuhua Zou (@ZouYuhua) <zouy@vmware.com>
+requirements:
+- "pycdlib"
+- "python >= 2.7"
+version_added: '5.4.0'
```
Comment (felixfontein) [inline plugins/modules/files/iso_get_files.py:22]: The 5.4.0 release is tomorrow and this will definitely not make it in that release (not enough time to properly review it otherwise), so feel free to bump the version to 5.5.0 already :-)

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution! I've added some first comments.

File: changelogs/fragments/5048-iso-customize-getfiles.yml (hunk 1 above)
Comment (ZouYuhua) [reply changelogs/fragments/5048-iso-customize-getfiles.yml]: Thanks. Done

File: tests/integration/targets/iso_customize/files/grub.cfg (hunk 2 above)
Comment (ZouYuhua) [reply tests/integration/targets/iso_customize/files/grub.cfg:1]: Thanks. Done

File: plugins/modules/files/iso_get_files.py (hunk 3 above)
Comment (ZouYuhua) [reply plugins/modules/files/iso_get_files.py]: Thanks. Done

File: plugins/modules/files/iso_customize.py [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+                    ret, msg = iso_delete_file(iso, item.strip())
+                    if ret != 0:
+                        MODULE_ISO_CUSTOMIZE.fail_json(msg=msg)
+            elif op == "add_files":
+                for item in data:
+                    ret, msg = iso_add_file(iso, item['src_file'],
+                                            item['dest_file'])
+                    if ret != 0:
+                        MODULE_ISO_CUSTOMIZE.fail_json(msg=msg)
+            elif op == "modify_files":
```
Comment (russoz) [inline plugins/modules/files/iso_customize.py:326]: I would recommend this feature be removed from this module:
* The file is being extracted and re-added for each entry - if multiple modifications are made on the same file, there are better ways to do that.
* If multiple modifications are done on the same file it is possible to alter the same line twice, which could cause a lot of confusion, and debugging *that* within this ISO module would be more challenging than what is needed.
* The code (on line 182) is assuming the file is in UTF-8, which may not be true. And having to deal with charsets in here is probably going to increase the complexity of the module beyond its purpose
* There is a number of different modules and filters to alter contents in Ansible and in this collection. Instead of creating yet another way to change files, it might be a better practice to leave that feature for the specialized tools, and make this module simpler.

File: plugins/modules/files/iso_customize.py [hunk 5]
```diff
@@ -0,0 +1,432 @@
...
+def get_local_tmp_dir():
+    dir_name = time.strftime('%Y-%m-%d-%H-%M-%S', time.localtime(time.time()))
+    tmp_dir = os.path.join('/tmp', dir_name)
+
+    try:
+        os.makedirs(tmp_dir)
+    except OSError as err:
+        msg = "Failed to create folder %s with error %s" % (tmp_dir, to_native(err))
+        MODULE_ISO_CUSTOMIZE.fail_json(msg=msg)
+
```
Comment (russoz) [inline plugins/modules/files/iso_customize.py:151]: You should use the functions in `tempfile` instead:
https://docs.python.org/3/library/tempfile.html

File: plugins/modules/files/iso_customize.py [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+        if record.is_file():
+            opened_iso.get_file_from_iso(file_local, rr_path=iso_file)
+        else:
+            return -1, "%s is not a file in ISO" % iso_file
+    except Exception as err:
+        msg = "Failed to get iso file %s with error: %s" % (iso_file, to_native(err))
+        MODULE_ISO_CUSTOMIZE.fail_json(msg=msg)
+
+    if not os.path.exists(file_local):
+        return -1, "Failed to get file %s in ISO. Does file exists in ISO ?" % iso_file
```
Comment (russoz) [inline plugins/modules/files/iso_customize.py:172]: I **strongly** suggest you raise an `Exception` when something goes wrong, instead of returning numerical error codes.

File: plugins/modules/files/iso_customize.py [hunk 7]
```diff
@@ -0,0 +1,432 @@
...
+
+def modify_local_file(local_file, regexp, replace):
+    if not os.path.exists(local_file):
+        return -1, "Not find file %s" % local_file
+
+    shutil.copy(local_file, "%s.org" % local_file)
+    with open(local_file, "r+", encoding="utf-8") as fp:
+        lines = fp.readlines()
+        for (i,) in enumerate(lines):
+            lines[i] = re.sub(regexp, replace, lines[i], count=0)
```
Comment (russoz) [inline plugins/modules/files/iso_customize.py:185]: This will be more pythonic as:
```suggestion
        regexp = re.compile(regexp)
        lines = [regexp.sub(replace, line, count=0) for line in lines]
```

File: plugins/modules/files/iso_customize.py [hunk 8]
```diff
@@ -0,0 +1,432 @@
...
+                return 0, ""
+
+            if parent_dir == "/":
+                current_dirpath = "/%s" % check_dirname
+            else:
+                current_dirpath = "%s/%s" % (parent_dir, check_dirname)
+
+            try:
+                opened_iso.add_directory(current_dirpath.upper(), rr_name=check_dirname)
+                break
```
Comment (russoz) [inline plugins/modules/files/iso_customize.py:214]: This would be the very last statement of the loop, no need to `break` then

File: plugins/modules/files/iso_get_files.py [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+requirements:
+- "pycdlib"
+- "python >= 2.7"
+version_added: '5.5.0'
+
+options:
+   iso:
+     description:
+     - This is the absolute paths of source ISO file.
+     - Will fail if specified ISO file does not exist on local machine.
```
Comment (russoz) [inline plugins/modules/files/iso_get_files.py:28]: I believe you mean the `target` machine. As Ansible have controller and target machines, saying "local" can be quite confusing. I have seen the term being used in more than one place here, possibly in both modules.

File: plugins/modules/files/iso_get_files.py [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+
+EXAMPLES = r'''
+- name: "Get file from ISO"
+  community.general.iso_get_files:
+    iso: "/Users/zouy/Documents/ubuntu-22.04-desktop-amd64.iso"
+    get_files:
+      - file_in_iso: "/preseed/ubuntu.seed"
+        file_local: "/tmp/ubuntu.seed.org"
+      - file_in_iso: "/boot/grub/grub.cfg"
+        file_local: "/tmp/grub.cfg.org"
```
Comment (russoz) [inline plugins/modules/files/iso_get_files.py:59]: This comment is a suggestion, rather than a request. I think life could be easier on the users if they could do something like:
```yaml
- name: "Get file from ISO"
  community.general.iso_get_files:
    iso: /Users/zouy/Documents/ubuntu-22.04-desktop-amd64.iso
    dest_dir: /tmp/iso_files
    get_files:
      - /preseed/ubuntu.seed
      - /boot/grub/grub.cfg
```
with the result being the retrieval of the files and storage at:
- `/tmp/iso_files/preseed/ubuntu.seed`
- `/tmp/iso_files/boot/grub/grub.cfg`

File: plugins/modules/files/iso_get_files.py [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+        iso.open(iso_path)
+
+        for item in get_files_list:
+            file_in_iso = item['file_in_iso'].strip()
+            file_local = item['file_local'].strip()
+
+            file_local_dir = os.path.dirname(file_local)
+            if not os.path.exists(file_local_dir):
+                try:
+                    os.makedirs(file_local_dir)
```
Comment (russoz) [inline plugins/modules/files/iso_get_files.py:109]: again, not a good idea

File: tests/integration/targets/iso_customize/tasks/test_customize_iso.yml [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+          - file: "/boot/grub/grub.cfg"
+            regexp: "timeout=30"
+            replace: "default=0\nset timeout=5"
+          - file: "/boot/grub/grub.cfg"
+            regexp: "/casper/vmlinuz file=/cdrom/preseed/ubuntu.seed maybe-ubiquity quiet splash ---"
+            replace: '/casper/vmlinuz file=/cdrom/preseed/ubuntu.seed boot=casper debug-ubiquity automatic-ubiquity quiet splash noprompt ---'
+      register: customize_iso_result
+
+    - name: "Display the ISO customized result"
+      debug: var=customize_iso_result
```
Comment (russoz) [inline tests/integration/targets/iso_customize/tasks/test_customize_iso.yml:31]: This test is not checking any assertions.

Please improve the test by actually checking for results.

File: plugins/modules/files/iso_get_files.py [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+        file_local: "/tmp/grub.cfg.org"
+'''
+
+RETURN = r'''
+iso:
+    description: paths of source ISO file..
+    returned: on success
+    type: str
+    sample: "/path/to/file.iso"
+file_in_iso:
```
Comment (russoz) [inline plugins/modules/files/iso_get_files.py:68]: Given that the input allow for multiple files, it is odd that this will return only one. Would that be the last one? The first one?

File: plugins/modules/files/iso_get_files.py [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+                    msg = "Failed to create folder %s with error: %s" % (
+                        file_local_dir, to_native(err))
+                    return -1, msg
+
+            record = iso.get_record(rr_path=file_in_iso)
+            if record.is_file():
+                iso.get_file_from_iso(file_local, rr_path=file_in_iso)
+            else:
+                iso.close()
+                return -1, "%s is not a file in ISO" % file_in_iso
```
Comment (russoz) [inline plugins/modules/files/iso_get_files.py:121]: This could be an option for the user: non-existent files should cause an error or should they be silently ignored?

**Review by russoz — CHANGES_REQUESTED** [review]
@ZouYuhua Thanks for your contribution! I left a number of comments as there are come changes that _must_ be done, other that _should_ and some that _could_ ;-)

File: plugins/modules/files/iso_customize.py [hunk 15]
```diff
@@ -0,0 +1,432 @@
...
+     required: no
+     suboptions:
+       src_file:
+         description:
+         - The absolute path with file name in local machine.
+         type: path
+       dest_file:
+         description:
+         - The absolute path with file name on ISO file.
+         type: path
```
Comment (felixfontein) [inline plugins/modules/files/iso_customize.py:56]: `path` is only a good type for files on the destination system. It should not be used for files anywhere else, since it makes sure that things like `$HOME` and `~` are expanded. Change this to `type: str`.

File: plugins/modules/files/iso_customize.py [hunk 16]
```diff
@@ -0,0 +1,432 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# Copyright: (c) 2022, Ansible Project
+# Copyright: (c) 2022, VMware, Inc. All Rights Reserved.
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```
Comment (felixfontein) [inline plugins/modules/files/iso_customize.py:6]:
```suggestion
# Copyright (c) 2022, Ansible Project
# Copyright (c) 2022, VMware, Inc. All Rights Reserved.
# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
# SPDX-License-Identifier: GPL-3.0-or-later
```

File: plugins/modules/files/iso_get_files.py [hunk 17]
```diff
@@ -0,0 +1,164 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# Copyright: (c) 2022, Ansible Project
+# Copyright: (c) 2022, VMware, Inc. All Rights Reserved.
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```
Comment (felixfontein) [inline plugins/modules/files/iso_get_files.py:6]:
```suggestion
# Copyright (c) 2022, Ansible Project
# Copyright (c) 2022, VMware, Inc. All Rights Reserved.
# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
# SPDX-License-Identifier: GPL-3.0-or-later
```

File: tests/integration/targets/iso_customize/aliases [hunk 18]
```diff
@@ -0,0 +1,4 @@
+shippable/posix/group1
```
Comment (felixfontein) [inline tests/integration/targets/iso_customize/aliases:1]:
```suggestion
# Copyright (c) 2022, Ansible Project
# Copyright (c) 2022, VMware, Inc. All Rights Reserved.
# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
# SPDX-License-Identifier: GPL-3.0-or-later

shippable/posix/group1
```

File: tests/integration/targets/iso_customize/files/grub.cfg (hunk 2 above)
Comment (felixfontein) [inline tests/integration/targets/iso_customize/files/grub.cfg:1]: (Assuming this file format allows comments.)
```suggestion
# Copyright (c) 2022, Ansible Project
# Copyright (c) 2022, VMware, Inc. All Rights Reserved.
# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
# SPDX-License-Identifier: GPL-3.0-or-later

set timeout=30
```

File: tests/integration/targets/iso_customize/files/ks.cfg [hunk 19]
```diff
@@ -0,0 +1,42 @@
+#Generated by Kickstart Configurator
```
Comment (felixfontein) [inline tests/integration/targets/iso_customize/files/ks.cfg:1]:
```suggestion
# Copyright (c) 2022, Ansible Project
# Copyright (c) 2022, VMware, Inc. All Rights Reserved.
# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
# SPDX-License-Identifier: GPL-3.0-or-later

#Generated by Kickstart Configurator
```

File: tests/integration/targets/iso_customize/files/ubuntu.seed [hunk 20]
```diff
@@ -0,0 +1,64 @@
+### step 1 : Localization
```
Comment (felixfontein) [inline tests/integration/targets/iso_customize/files/ubuntu.seed:1]:
```suggestion
# Copyright (c) 2022, Ansible Project
# Copyright (c) 2022, VMware, Inc. All Rights Reserved.
# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
# SPDX-License-Identifier: GPL-3.0-or-later

### step 1 : Localization
```

File: tests/integration/targets/iso_customize/files/ks.cfg [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+auth  --useshadow  --enablemd5 
+#Network information
+network --bootproto=dhcp --device=eth0
+#Firewall configuration
+firewall --disabled --ssh 
+#Do not configure the X Window System
+skipx
+#install packages
+%packages
+openssh-server
```
Comment (felixfontein) [inline tests/integration/targets/iso_customize/files/ks.cfg:42]: Since it is not possible to add comments to the binary files below: could you please remove them, and instead add some tasks at the beginning of the tests which create them?

File: tests/integration/targets/iso_customize/tasks/test_customize_iso.yml [hunk 22]
```diff
@@ -0,0 +1,31 @@
+- name: iso_customize
```
Comment (felixfontein) [inline tests/integration/targets/iso_customize/tasks/test_customize_iso.yml:1]:
```suggestion
---
# Copyright (c) 2022, Ansible Project
# Copyright (c) 2022, VMware, Inc. All Rights Reserved.
# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
# SPDX-License-Identifier: GPL-3.0-or-later

- name: iso_customize
```

File: tests/integration/targets/iso_get_files/aliases [hunk 23]
```diff
@@ -0,0 +1,4 @@
+shippable/posix/group1
```
Comment (felixfontein) [inline tests/integration/targets/iso_get_files/aliases:1]:
```suggestion
# Copyright (c) 2022, Ansible Project
# Copyright (c) 2022, VMware, Inc. All Rights Reserved.
# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
# SPDX-License-Identifier: GPL-3.0-or-later

shippable/posix/group1
```

File: tests/integration/targets/iso_get_files/tasks/test_get_iso_files.yml [hunk 24]
```diff
@@ -0,0 +1,18 @@
+- name: iso_get_files
```
Comment (felixfontein) [inline tests/integration/targets/iso_get_files/tasks/test_get_iso_files.yml:1]:
```suggestion
---
# Copyright (c) 2022, Ansible Project
# Copyright (c) 2022, VMware, Inc. All Rights Reserved.
# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
# SPDX-License-Identifier: GPL-3.0-or-later

- name: iso_get_files
```

File: tests/integration/targets/iso_get_files/tasks/test_get_iso_files.yml [hunk 25]
```diff
@@ -0,0 +1,18 @@
...
+        iso: "{{ test_files_dir }}/test.iso"
+        get_files:
+          - file_in_iso: "/1.txt"
+            file_local: "/tmp/aaa/1.txt.org"
+          - file_in_iso: "/delete/1.txt"
+            file_local: "/tmp/aaa/2.txt"
+      register: iso_get_files
+
+    - name: "Display the ISO customized result"
+      debug: var=iso_get_files
```
Comment (felixfontein) [inline tests/integration/targets/iso_get_files/tasks/test_get_iso_files.yml:18]: Please make sure that all text files have a trailing newline.

File: plugins/modules/files/iso_customize.py [hunk 26]
```diff
@@ -0,0 +1,432 @@
...
+DOCUMENTATION = r'''
+---
+module: iso_customize
+short_description: Customize the ISO file by add/delete/modify files
+description:
+    - This module is used to customize the ISO file by add/delete/modify files.
+author:
+    - Yuhua Zou (@ZouYuhua) <zouy@vmware.com>
+requirements:
+- "pycdlib"
```
[the remark this answers is not in the document]
Comment (ZouYuhua) [reply plugins/modules/files/iso_customize.py]: Thanks. Done

File: plugins/modules/files/iso_customize.py [hunk 27]
```diff
@@ -0,0 +1,432 @@
...
+        module.fail_json(
+            missing_required_lib('pycdlib'), exception=PYCDLIB_IMP_ERR)
+
+    src_iso = module.params.get('src_iso')
+    if not os.path.exists(src_iso):
+        module.fail_json(msg="The %s does not exist." % src_iso)
+
+    dest_iso = module.params.get('dest_iso')
+    if dest_iso and len(dest_iso) == 0:
+        module.fail_json(str(msg='Please specify the absolute path of the customized ISO file \
```
[the remark this answers is not in the document]
Comment (ZouYuhua) [reply plugins/modules/files/iso_customize.py]: Thanks. Done

File: plugins/modules/files/iso_customize.py [hunk 28]
```diff
@@ -0,0 +1,432 @@
...
+    if add_files_list and len(add_files_list) > 0:
+        for item in add_files_list:
+            if not os.path.exists(item['src_file']):
+                module.fail_json(msg="The %s does not exist." % item['src_file'])
+        op_data_item['op'] = "add_files"
+        op_data_item['data'] = add_files_list
+        op_data_list.append(copy.deepcopy(op_data_item))
+
+    modify_files_list = module.params.get('modify_files')
+    if modify_files_list and len(modify_files_list) > 0:
```
[the remark this answers is not in the document]
Comment (ZouYuhua) [reply plugins/modules/files/iso_customize.py]: Thanks. changed

File: plugins/modules/files/iso_get_files.py (hunk 10 above)
Comment (ZouYuhua) [reply plugins/modules/files/iso_get_files.py]: Thanks. It's very helpful.

File: plugins/modules/files/iso_get_files.py (hunk 13 above)
Comment (ZouYuhua) [reply plugins/modules/files/iso_get_files.py]: Thanks. It's changed

File: plugins/modules/files/iso_customize.py (hunk 16 above)
Comment (ZouYuhua) [reply plugins/modules/files/iso_customize.py]: Thanks. It's changed

File: plugins/modules/files/iso_get_files.py (hunk 17 above)
Comment (ZouYuhua) [reply plugins/modules/files/iso_get_files.py]: Thanks. It's changed

File: tests/integration/targets/iso_customize/aliases (hunk 18 above)
Comment (ZouYuhua) [reply tests/integration/targets/iso_customize/aliases:6]: Thanks. It's changed

File: tests/integration/targets/iso_customize/files/grub.cfg (hunk 2 above)
Comment (ZouYuhua) [reply tests/integration/targets/iso_customize/files/grub.cfg:1]: Thanks. Please ignore this file.

File: tests/integration/targets/iso_customize/files/ks.cfg (hunk 21 above)
Comment (ZouYuhua) [reply tests/integration/targets/iso_customize/files/ks.cfg]: Thanks. The file is deleted.

File: tests/integration/targets/iso_customize/tasks/test_customize_iso.yml (hunk 22 above)
Comment (ZouYuhua) [reply tests/integration/targets/iso_customize/tasks/test_customize_iso.yml:6]: Thanks. It's changed

File: tests/integration/targets/iso_get_files/tasks/test_get_iso_files.yml (hunk 24 above)
Comment (ZouYuhua) [reply tests/integration/targets/iso_get_files/tasks/test_get_iso_files.yml:6]: Thanks. It's changed

File: tests/integration/targets/iso_get_files/tasks/test_get_iso_files.yml (hunk 25 above)
Comment (ZouYuhua) [reply tests/integration/targets/iso_get_files/tasks/test_get_iso_files.yml]: Thanks. It's changed

File: tests/integration/targets/iso_get_files/aliases (hunk 23 above)
Comment (ZouYuhua) [reply tests/integration/targets/iso_get_files/aliases:6]: Thanks. It's changed

File: tests/integration/targets/iso_customize/files/ks.cfg (hunk 19 above)
Comment (ZouYuhua) [reply tests/integration/targets/iso_customize/files/ks.cfg]: Thanks. The file is deleted.

File: tests/integration/targets/iso_customize/tasks/test_customize_iso.yml (hunk 12 above)
Comment (ZouYuhua) [reply tests/integration/targets/iso_customize/tasks/test_customize_iso.yml]: Thanks. It's changed

File: plugins/modules/files/iso_customize.py (hunk 5 above)
Comment (ZouYuhua) [reply plugins/modules/files/iso_customize.py]: Thanks. It's changed.

File: plugins/modules/files/iso_customize.py [hunk 29]
```diff
@@ -0,0 +1,432 @@
...
+    dest_iso = module.params.get('dest_iso')
+    if dest_iso and len(dest_iso) == 0:
+        module.fail_json(str(msg='Please specify the absolute path of the customized ISO file \
+            using dest_iso parameter.'))
+
+    dest_iso_dir = os.path.dirname(dest_iso)
+    if dest_iso_dir and not os.path.exists(dest_iso_dir):
+        # will create intermediate dir for customized ISO file
+        try:
+            os.makedirs(dest_iso_dir)
```
[the remark this answers is not in the document]
Comment (ZouYuhua) [reply plugins/modules/files/iso_customize.py:376]: Thanks. It's changed

File: plugins/modules/files/iso_get_files.py (hunk 14 above)
Comment (ZouYuhua) [reply plugins/modules/files/iso_get_files.py:121]: Thanks for your code review.
 We have checked the file exists when invoke iso.get_file_from_iso(file_local, rr_path=file_in_iso) :
 record = iso.get_record(rr_path=file_in_iso)
 if record.is_file():
       iso.get_file_from_iso(file_local, rr_path=file_in_iso)

File: plugins/modules/files/iso_get_files.py [hunk 30]
```diff
@@ -0,0 +1,170 @@
...
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+DOCUMENTATION = r'''
+---
+module: iso_get_files
+short_description: Fetch files from ISO
+description:
+  - This module is used to get specific files from ISO.
```
Comment (felixfontein) [inline plugins/modules/files/iso_get_files.py:17]: What I'm not sure is: how is this module related to the existing `iso_extract` module?

File: plugins/modules/files/iso_customize.py (hunk 4 above)
Comment (ZouYuhua) [reply plugins/modules/files/iso_customize.py]: Thanks. remove this feature.

File: plugins/modules/files/iso_customize.py (hunk 15 above)
Comment (ZouYuhua) [reply plugins/modules/files/iso_customize.py]: Thanks. changed

File: plugins/modules/files/iso_get_files.py (hunk 11 above)
Comment (ZouYuhua) [reply plugins/modules/files/iso_get_files.py:109]: Using try-exception-errno to avoid thread security issue will be available workaround currently.

1) mkdtemp won't works well here: we need to create directory and subdirectory with given path.
2) mkpath: [Bug 10948](http://bugs.python.org/issue10948)
3) library pathlib:/pathlib2: we need to install the package for old python.

I find many other files also use function "makedirs(path)".  We can change it when find the standard solution.


Best regards
Yuhua Zou

File: plugins/modules/files/iso_get_files.py (hunk 30 above)
Comment (ZouYuhua) [reply plugins/modules/files/iso_get_files.py:17]: iso_extract will extract the whole iso content. It will consume time and space if the ISO have big size.

Background:
For some OS ISO, we need to modify the the original iso in order to support unatended installation
the easiest way is to modify the original iso with API but not "extract the ISO, modify file, regenerate new ISO"
 
Thanks
Yuhua Zou

Comment (felixfontein) [reply plugins/modules/files/iso_get_files.py:17]: From 'iso_extract's documentation, it requires a list of files to be specified that will be extracted. It does not seem to extract the whole ISO. How is this different from your module?

**Comment by ZouYuhua** [comment]
@felixfontein @russoz 
Hi Admin,
Anything I need to do to for the error "Verify REUSE / check (pull_request)  ?   Thanks very much

Best regards
Yuhua Zou

**Comment by felixfontein** [comment]
> @felixfontein @russoz Hi Admin, Anything I need to do to for the error "Verify REUSE / check (pull_request) ? Thanks very much

Yes. As long as that test does not pass, this cannot be merged.

Also please note that we do not accept new binary files for tests, except in very rare circumstances. Please use the `iso_create` module to create the `.iso` files you need during the tests at the beginning of your tests, instead of simply including the full `.iso` files with this PR.

**Comment by ZouYuhua** [comment]
@felixfontein @russoz
Hi Admin,
I don't know how to fix these two issues. 

1. CI (Sanity devel Test extra)
01:10 .github/BOTMETA.yml:0:0: Author ZouYuhua not mentioned as active or inactive maintainer for plugins/modules/files/iso_customize.py (mentioned are: )
01:10 .github/BOTMETA.yml:0:0: No (active or inactive) maintainer mentioned for plugins/modules/files/iso_customize.py


2. CI (Docker devel Ubuntu 20.04 -1)
fatal: [testhost]: FAILED! => {
07:33     "changed": false,
07:33     "cmd": "/usr/sbin/mkfs.ocfs2 -Fx /tmp/ansible.n2bd9l6q.test/img",
07:33     "invocation": {
07:33         "module_args": {
07:33             "dev": "/tmp/ansible.n2bd9l6q.test/img",
07:33             "force": false,
07:33             "fstype": "ocfs2",
07:33             "opts": null,
07:33             "resizefs": false,
07:33             "state": "present"
07:33         }
07:33     },
07:33     "msg": "mkfs.ocfs2 1.8.6\nmkfs.ocfs2: Could not open device /tmp/ansible.n2bd9l6q.test/img: Invalid argument\nmkfs.ocfs2: Could not open device /tmp/ansible.n2bd9l6q.test/img: Invalid argument\nmkfs.ocfs2: Could not open device /tmp/ansible.n2bd9l6q.test/img: Invalid argument\nmkfs.ocfs2: Could not open device /tmp/ansible.n2bd9l6q.test/img: Invalid argument",
07:33     "rc": 1,
07:33     "stderr": "mkfs.ocfs2 1.8.6\nmkfs.ocfs2: Could not open device /tmp/ansible.n2bd9l6q.test/img: Invalid argument\nmkfs.ocfs2: Could not open device /tmp/ansible.n2bd9l6q.test/img: Invalid argument\nmkfs.ocfs2: Could not open device /tmp/ansible.n2bd9l6q.test/img: Invalid argument\nmkfs.ocfs2: Could not open device /tmp/ansible.n2bd9l6q.test/img: Invalid argument\n",
07:33     "stderr_lines": [
07:33         "mkfs.ocfs2 1.8.6",
07:33         "mkfs.ocfs2: Could not open device /tmp/ansible.n2bd9l6q.test/img: Invalid argument",
07:33         "mkfs.ocfs2: Could not open device /tmp/ansible.n2bd9l6q.test/img: Invalid argument",
07:33         "mkfs.ocfs2: Could not open device /tmp/ansible.n2bd9l6q.test/img: Invalid argument",
07:33         "mkfs.ocfs2: Could not open device /tmp/ansible.n2bd9l6q.test/img: Invalid argument"
07:33     ],
07:33     "stdout": "Cluster stack: classic o2cb\nCluster stack: classic o2cb\nCluster stack: classic o2cb\nCluster stack: classic o2cb\n",
07:33     "stdout_lines": [
07:33         "Cluster stack: classic o2cb",
07:33         "Cluster stack: classic o2cb",
07:33         "Cluster stack: classic o2cb",
07:33         "Cluster stack: classic o2cb"
07:33     ]
07:33 }

It seems I don't have the privilege to rerun the failed job. 
Any suggestion ?   Thanks very much.  It seems I don't have the privilege to rerun the failed job.


Best regards
Yuhua Zou

**Comment by ZouYuhua** [comment]
@felixfontein @russoz 
I have one failure with CI test and don't know how to fix it:
  Test "botmeta" failed with the following 2 errors:
  01:05 .github/BOTMETA.yml:0:0: Author ZouYuhua not mentioned as active or inactive maintainer for plugins/modules/files  /iso_customize.py (mentioned are: )
 01:05 .github/BOTMETA.yml:0:0: No (active or inactive) maintainer mentioned for plugins/modules/files/iso_customize.py

Would you please help me with it ?   Thanks very much.

Best regards
Yuhua Zou

**Comment by russoz** [comment]
Hi @ZouYuhua 

You need to review https://github.com/ansible-collections/community.general/blob/main/CONTRIBUTING.md#creating-new-modules-or-plugins carefully. You are missing step #5 there.

But notice that you have merge commits piling up there. I would recommend working on those first. The bot message contains a link to the relevant documentation.

**Comment by ZouYuhua** [comment]
I open a new pull request https://github.com/ansible-collections/community.general/pull/5190.
